verilog/analysis/flow_tree.h

```diff
@@ -0,0 +1,114 @@
+#include "common/lexer/token_stream_adapter.h"
+#include "verilog/parser/verilog_token_enum.h"
+
+namespace verilog {
+using TokenSequenceConstIterator = verible::TokenSequence::const_iterator;
+
+// Receive a complete token sequence of one variant.
+// 1st argument: generated variant.
+// 2nd argument: number of generated variants (to decide if more is needed).
+// 3rd argument: checks if the variant is ready to be sent.
+using VariantReceiver =
+    std::function<bool(const verible::TokenSequence &, const int, const bool)>;
+
+struct ConditionalBlock {
+  TokenSequenceConstIterator ifdef_iterator;
+  TokenSequenceConstIterator ifndef_iterator;
+  std::vector<TokenSequenceConstIterator> elsif_iterators;
+  TokenSequenceConstIterator else_iterator;
+  TokenSequenceConstIterator endif_iterator;
+};
+
+// FlowTree class builds the control flow graph of a tokenized System-Verilog
+// source code. Furthermore, enabling doing the following queries on the graph:
+// 1) Generating all the possible variants (provided via a callback function).
+// 2) TODO(karimtera): uniquely identify a variant with a bitset.
+class FlowTree {
+ public:
+  explicit FlowTree(verible::TokenSequence source_sequence)
+      : source_sequence_(std::move(source_sequence)){};
+
+  // Constructs the control flow tree by adding the tree edges in edges_.
+  absl::Status GenerateControlFlowTree();
+
+  // Generates all possible variants.
+  absl::Status GenerateVariants(const VariantReceiver &receiver);
+
+ private:
+  // Traveses the tree in a depth first manner.
+  absl::Status DepthFirstSearch(const VariantReceiver &receiver,
+                                TokenSequenceConstIterator current_node,
+                                std::bitset<128> assumed);
+
+  // Checks if the iterator points to a conditonal directive (`ifdef/ifndef...).
+  static bool IsConditional(TokenSequenceConstIterator iterator);
+
+  //
+  absl::Status AddBlockEdges(const ConditionalBlock &block);
+
+  // The tree edges which defines the possible next childs of each token in
+  // source_sequence_.
+  std::map<TokenSequenceConstIterator, std::vector<TokenSequenceConstIterator>>
+      edges_;
+
+  // Extracts the conditional macro checked.
+  static absl::Status MacroFollows(
+      TokenSequenceConstIterator conditional_iterator);
+
+  // Adds macro to conditional_macro_id_ map.
+  absl::Status AddMacroOfConditionalToMap(
+      TokenSequenceConstIterator conditional_iterator);
+
+  int GetMacroIDOfConditional(TokenSequenceConstIterator conditional_iterator);
+
+  // Holds all of the conditional blocks.
+  std::vector<ConditionalBlock> if_blocks_;
+
+  // The original source code lexed token seqouence.
+  const verible::TokenSequence source_sequence_;
+
+  // The variant's token sequence currrently being built by DepthFirstSearch.
+  verible::TokenSequence current_sequence_;
+
+  // Macros assumed in the variant that is currently begin built.
+  // TODO(karimtera): Is it better to use dymanic_bitset?
```

Comment:
or a vector in which we store `uints_64` and do some bid-fiddeling.
For now, this is fine, but make sure in a unit-test that if you generate more than 128 different macros, that the code reacts gracefully, e.g. by returning a Status in which it returns a failure.

Anyway, since we use the same type in other places as well (e.g. in the parameter to DepthFirst search), I suggest to make it a using declaration/typedef. The common place to define it would be right after `'private:`

```c++
private:
  using BitSet = std::bitset<128>;
  // ...
  absl::Status DepthFirstSearch(..., BitSet,...);
  //...
  BitSet current_macros_;
```